packages/webpack-inline-constant-exports-plugin/index.js

```
@@ -0,0 +1,117 @@
+							if ( declarator.init.type === 'Literal' ) {
+								addConstantExport( module, declarator.id.name, declarator.init.raw );
+							}
+						}
+					}
+
+					if ( statement.type === 'ExportDefaultDeclaration' && declaration.type === 'Literal' ) {
+						addConstantExport( module, 'default', declaration.raw );
+					}
+				};
+
+				parser.hooks.exportDeclaration.tap( 'InlinePlugin', handleExport );
+				parser.hooks.exportExpression.tap( 'InlinePlugin', handleExport );
+			};
+
+			moduleTypes.forEach( moduleType => {
+				normalModuleFactory.hooks.parser.for( moduleType ).tap( 'InlinePlugin', handleParser );
+			} );
+
+			compilation.hooks.optimizeDependencies.tap( 'InlinePlugin', modules => {
+				for ( const module of modules ) {
+					const importStatementDependencies = new Map();
+					const usesNonConstantDependencies = new Map();
+					const importSpecifierDependencies = [];
+
+					for ( const dep of module.dependencies ) {
+						if ( ! this.isConstantsModule( dep.module ) ) {
+							continue;
+						}
+
+						if ( dep instanceof HarmonyImportSideEffectDependency ) {
+							importStatementDependencies.set( dep.module, dep );
+						}
+
+						if ( dep instanceof HarmonyImportSpecifierDependency ) {
+							if ( dep.module.constantExports && dep.id in dep.module.constantExports ) {
+								importSpecifierDependencies.push( dep );
+								const inlineDep = new ConstDependency(
+									[
+										'/* <inline> */ ',
+										dep.shorthand ? `${ dep.name }: ` : '',
+										dep.module.constantExports[ dep.id ],
+										' /* </inline> */',
+									].join( '' ),
+									dep.range
+								);
+								inlineDep.loc = dep.loc;
+								module.addDependency( inlineDep );
+							} else {
+								usesNonConstantDependencies.set( dep.module, true );
+							}
+						}
+					}
+
+					for ( const [ depModule, depImport ] of importStatementDependencies ) {
+						if ( ! usesNonConstantDependencies.get( depModule ) ) {
```

This is a fallback for when a module with something that isn't a constant has been incorrectly added, correct? If so, and if modules with non-constants aren't meant to be used with this plugin, should we perhaps emit a warning instead of silently skipping the optimisation?

---

as a general plugin I think we have to leave the dependency in cases where we import constants _and_ non-constants, right?

considering all the questions we all have here, I think we need to both applaud @jsnajdr and get higher-level intentionality documented throughout the code